### ISSUE TITLE
[TySan] false positive with unions and placement-new

### ISSUE THREAD

**Issue by Jannik2099** [llvm/llvm-project#120987]
According to https://eel.is/c++draft/class.union#general-6 , the active union member in an union of non-POD types can be switched by invoking the dtor of the old member, and placement new-ing into the new member.

However, tysan does not seem to like this https://godbolt.org/z/nvnjzjGez

```cpp
#include <cstdio>
#include <new>

class Foo {
   private:
    short i{};

   public:
    Foo() { printf("Foo nontrivial ctor\n"); }
    ~Foo() { printf("Foo nontrivial dtor\n"); }
};

class Bar {
   private:
    int i{};

   public:
    Bar() { printf("Bar nontrivial ctor\n"); }
    ~Bar() { printf("Bar nontrivial dtor\n"); }
};

union U {
    Foo foo;
    Bar bar;
    ~U() {}
};

int main() { 
    U u{ .foo = {}};
    u.foo.~Foo();
    new (&u.bar) Bar;
}
```

```
==1==ERROR: TypeSanitizer: type-aliasing-violation on address 0x7ffff9117e2c (pc 0x61eed1b9b89d bp 0x7ffff9117d10 sp 0x7ffff9117cb8 tid 1)
WRITE of size 4 at 0x7ffff9117e2c with type int (in Bar at offset 0) accesses an existing object of type short (in Foo at offset 0)
    #0 0x61eed1b9b89c  (/app/output.s+0x2b89c)
```

This also happens with an union of POD or trivial types, but I'm not 100% sure on the standard working if placement-new is supposed to be allowed in these cases.

**Comment by gbMattN** [comment]
TySan currently doesn't get the correct type information from unions, [this is a more detailed explaination](https://discourse.llvm.org/t/reviving-typesanitizer-a-sanitizer-to-catch-type-based-aliasing-violations/66092/14?u=gbmattn).

I plan to add something to let it understand types TBAA metadata doesn't currently cover, but I don't have anything yet.

**Comment by zygoloid** [comment]
The same bug happens without a union too: https://godbolt.org/z/rncen8Pe8

**Comment by gbMattN** [comment]
```
new (&u.data) Foo{};
new (&u.data) Bar{};
```
In this case you are using the same address to initialize the foo and bar. I'm pretty sure that _is_ and alias violation?

Also regarding my prior comment about adding something, it turned out to be a lot more complicated than I first thought so it is on the backburner for now 😅

**Comment by zygoloid** [comment]
> ```
> new (&u.data) Foo{};
> new (&u.data) Bar{};
> ```
> 
> In this case you are using the same address to initialize the foo and bar. I'm pretty sure that _is_ and alias violation?

Nope. The C++ rules say that a `new` expression here creates a new `Bar` object and destroys the old `Foo` object, so the `new` expression should have the effect of changing the type of the storage from `Foo` to `Bar` for aliasing purposes.

**Comment by zygoloid** [comment]
Here's an even simpler C example [with a false positive](https://godbolt.org/z/q3P13zh9q):
```c
#include <stdlib.h>
int main() { 
  void *p = malloc(4);
  *(short*)p = 0;
  *(int*)p = 0;
}
```
The C rules and the C++ rules are different here: in C++, the above *would* be invalid, because the type of memory is only changed by a few special operations (new-expressions, assignment to union members, memcpy, a handful of others). But in C, the type of memory that doesn't have a declared type (for example, heap memory) can be changed simply by assigning to it.

**Comment by gbMattN** [comment]
LLVM's tbaa metadata is incorrect in this regard. This has been discussed in https://github.com/llvm/llvm-project/issues/54878 (and another issue I don't have to hand right now).

LLVM's tbaa is static per object, wheras in both C and C++ different operations can change the underlying type. This is not something that can be fixed on the TypeSanitizer side, since it technically _is_ correctly detecting what LLVM sees as an alias violation.

I don't see this getting fixed imminently

**Comment by gbMattN** [comment]
Other issue -> https://github.com/llvm/llvm-project/issues/58225
